### PR TITLE
Disable incompatible Snacks, use correct URLs

### PR DESCRIPTION
**Pull request by brentvatne**
# Why

- We should let people know before even trying to launch a Snack that it won't work in their version of Expo Go, based on the SDK version of the Snack and of Expo Go
- We need to use the new Snack URLs as generated by the snack-content package or you can't open any saved Snacks!

# How

image

image

Logged an example URL:

```
exp://u.expo.dev/933fd9c0-1666-11e7-afca-d980795c5824?runtime-version=exposdk%3A52.0.0&channel-name=production&snack=%40brents%2Fcourageous-violet-cookies
```

Looks right?

### CODE REVIEW

File: apps/expo-go/src/components/SnacksListItem.tsx [hunk 1]
```diff
@@ -24,43 +27,60 @@ function normalizeDescription(description?: string): string | undefined {
...
+  const { fullName, name, description, isDraft, first, last, sdkVersion } = snackData;
   const theme = useExpoTheme();
 
-  const handlePressProject = () => {
-    Linking.openURL(UrlUtils.normalizeUrl(url));
-  };
+  const normalizedDescription = normalizeDescription(description);
+  const isSupported = sdkVersion ? sdkVersion === Environment.supportedSdksString : true;
 
-  const handleLongPressProject = () => {
```
Comment (brentvatne) [inline apps/expo-go/src/components/SnacksListItem.tsx:34]: got rid of the share dialog, not very useful

File: apps/expo-go/src/components/SnacksListItem.tsx [hunk 2]
```diff
@@ -24,43 +27,60 @@ function normalizeDescription(description?: string): string | undefined {
...
           roundedTop={first ? 'large' : undefined}
           roundedBottom={last ? 'large' : undefined}>
           <Row align="center" justify="between">
             <View align="start" flex="1">
-              <Text type="InterSemiBold" ellipsizeMode="tail" numberOfLines={1}>
+              <Text
+                type="InterSemiBold"
+                ellipsizeMode="tail"
+                numberOfLines={1}
+                style={{ opacity: isSupported ? 1 : 0.5 }}>
```
Comment (brentvatne) [inline apps/expo-go/src/components/SnacksListItem.tsx:83]: i tried making the entire row lower opacity but it felt like a bit much, the text + the badge communicates sufficiently imo

File: apps/expo-go/src/components/SnacksListItem.tsx [hunk 3]
```diff
@@ -71,6 +91,16 @@ export function SnacksListItem({ description, isDraft, name, url, first, last }:
                   </Text>
                 </>
               )}
+              {!isSupported && (
+                <>
+                  <Spacer.Vertical size="tiny" />
+                  <View bg="secondary" rounded="medium" flex="0" padding="tiny" border="default">
+                    <Text size="small" type="InterRegular">
+                      Unsupported SDK ({sdkVersion})
```
Comment (brentvatne) [inline apps/expo-go/src/components/SnacksListItem.tsx:99]: "the badge" referenced above

File: apps/expo-go/src/navigation/Navigation.tsx [hunk 4]
```diff
@@ -235,17 +236,18 @@ export default (props: { theme: ColorTheme }) => {
...
-                        .routes.findIndex((r: RouteProp<any, any>) => r.key === route.key) > 0
-                        ? 0
-                        : undefined,
+                    // NOTE(brentvatne): it is unclear what this was intended for, it doesn't appear to be needed?
+                    // headerStatusBarHeight:
+                    //   navigation
+                    //     .getState()
+                    //     .routes.findIndex((r: RouteProp<any, any>) => r.key === route.key) > 0
+                    //     ? 0
+                    //     : undefined,
```
Comment (brentvatne) [inline apps/expo-go/src/navigation/Navigation.tsx:250]: kind of unrelated to this PR but i couldn't figure out what this was for and removing it appeared to (perhaps by chance!) fix some visual glitches in the header

File: apps/expo-go/src/screens/HomeScreen/SnacksSection.tsx [hunk 5]
```diff
@@ -32,10 +32,12 @@ export function SnacksSection({ snacks, showMore, accountName }: Props) {
         return (
           <Fragment key={snack.id}>
             <SnacksListItem
+              id={snack.id}
               name={snack.name}
+              sdkVersion={snack.sdkVersion}
               description={snack.description}
               isDraft={snack.isDraft}
-              url={snack.fullName}
+              fullName={snack.fullName}
```
Comment (brentvatne) [inline apps/expo-go/src/screens/HomeScreen/SnacksSection.tsx:40]: renamed from url because it's.. not a url

**Review by alanjhughes — APPROVED** [review]

File: apps/expo-go/src/utils/UrlUtils.ts [hunk 6]
```diff
@@ -1,9 +1,20 @@
...
 import url from 'url';
 
+import Environment from './Environment';
 import Config from '../api/Config';
 
 const HTTPS_HOSTS = [Config.api.host, 'exp.host', 'exponentjs.com', 'getexponent.com'];
 
+export function normalizeSnackUrl(fullName: string, channelName?: string): string {
+  return createRuntimeUrl({
+    // Only ever try to open Snack using the current SDK version
```
Comment (byCedric) [inline apps/expo-go/src/utils/UrlUtils.ts:11]: No worries if you don't do this btw, basically users will run into the "This SDK version is not supported" screen. So it's still a useful/actionable error.

**Review by byCedric — APPROVED** [review]

**Review by wschurman — APPROVED** [review]

**Pull request merged**